### ISSUE TITLE
Gridded data and Netcdf

### ISSUE THREAD

**Issue by rdstern** [africanmathsinitiative/R-Instat#3995]
This dialogue **Climatic > File > Import Gridded Data (IRI)** seems to work - but at closer inspection it has serious problems.
(See the next comment before trying to correct them.)
 
1. I assume it needs an internet connection and we should perhaps mention that?
2. The up-down controls should be updated to our new version, so it is easier to type into them.  And the up-down steps should be 0.005 and not 0.0001.  (I assume this should be quite easy to do.)
3. It has its own waiting dialogue, though that isn't very useful.  Perhaps it should be replaced by information on how many rows of data will be imported, so we can be cautious if there are too many.  But this would be for later.
4. The date column is imported as a factor.  I presume could be a date column.  It can be made into a date column, though the format (%d<space>%b<space>%Y) could usefully be added to our standard ones in the Make Date dialogue.
5. Now the big one.  It only imports part of the data.  When I imported one point, it was from 1981 to 1987.  With an area (500 points) it was less than 2 years.  I don't know if this was to do with my internet connection and would like to try with a better one.  I don't know whether the IRI site allows a choice of first and last date?  

It does seem to import sensibly into 2 data frames.  I would like to use it in the forthcoming training in Guyana.

**Comment by rdstern** [comment]
The Climatic > File menu has special dialogues for SSTs (Sea Surface temperatures) and for importing gridded data from IRI - see comment above.  
We also now have an option to import netCDF files.
I now compare them, and wonder if we still need the dialogue for the gridded IRI data.
That dialogue goes (I think) to the IRI site.  It imports data from either a point, or an area from that site.  I like that option, particularly for a point.  I wonder how that option works.
 An alternative could be to make it a 2-step process.  This is particularly because the gridded data doesn't seem to work well yet.
So I went to the IRI site.  Here are the data from CHIRPS in  the form we need, i.e. time series.  If you go to the CHIRPS site you get a separate file per month or year. 
[https://iridl.ldeo.columbia.edu/SOURCES/.UCSB/.CHIRPS/.v2p0/.daily-improved/.global/.0p05/.prcp/datafiles.html](url)   
On this site you can download all the data (for the whole world) in a variety of formats.  But not in netCDF, because it is too large.  But there is an option on this page to select a subset of the data.  So I did. For a 1 degree square (about) in Guyana, that I had tried unsuccessfully with the gridded data.


I then tried to open it as a netcdf file.
And it worked.  Well, sort of.  It imported the whole file.  Hooray.  But it didn't have the details to allow me to get a subset, nor to provide the date in the correct format.

So, except for the option (which is useful) to download data from a single point.  I am left to wonder whether we need the dialogue for the gridded data.  And then I assume we also get the SSTs from the IRI site.  So will we need that dialogue either?

**Comment by rdstern** [comment]
I have now looked (in an editor) at the netCDF file.  Danny was of the opinion that it was just the specific column names (X, Y, T) that were not recognised in the dialogue.  Hence it was not possible to choose a subset.  If I could choose one, then I would like to download the data for the whole of Guyana (if allowed with the netcdf format.)  And then a user could choose the point or region they would like when they input the data into R-Instat.
In the top line I saw the X,Y and T.  I wonder if I just need to change those?  Or at least that would be one option?

**Comment by dannyparsons** [comment]
At this point I think we can only disable **Import Gridded Data (IRI)** if it's not working well as it will take longer to fix. If it is useful we can leave it as is.

**Comment by shadrackkibet** [comment]
This is just to report on the progress so far. It is not yet working properly therefore I am working on fixing bugs. I inherited URLs from existing R code which seems to need updating to work properly.

**Designer update:**

![image](https://user-images.githubusercontent.com/26301973/91887543-93120200-ec93-11ea-8f6c-29140fdc72f8.png)
![image](https://user-images.githubusercontent.com/26301973/91887565-9c9b6a00-ec93-11ea-9258-ca90b967cf45.png)
![image](https://user-images.githubusercontent.com/26301973/91887601-a91fc280-ec93-11ea-8893-99b82d778bba.png)
![image](https://user-images.githubusercontent.com/26301973/91887639-b472ee00-ec93-11ea-86a5-666b20cb84bb.png)
![image](https://user-images.githubusercontent.com/26301973/91887718-d0768f80-ec93-11ea-9aac-a8e3cd38e6b1.png)

**R code**
```

add_xy_area_range <- function(path, min_lon, max_lon, min_lat, max_lat) {
  paste0(path, "/X", "/",
         "(", ifelse(min_lon < 0, paste0(abs(min_lon), "W"), paste0(min_lon, "E")), ")", "/",
         "(", ifelse(max_lon < 0, paste0(abs(max_lon), "W"), paste0(max_lon, "E")), ")", "/",
         "RANGEEDGES", "/",
         "Y", "/",
         "(", ifelse(min_lat < 0, paste0(abs(min_lat), "S"), paste0(min_lat, "N")), ")", "/",
         "(", ifelse(max_lat < 0, paste0(abs(max_lat), "S"), paste0(max_lat, "N")), ")", "/",
         "RANGEEDGES", "/")
}

add_xy_point_range <- function(path, min_lon, min_lat) {
  paste0(path, "/X", "/",
         "(", ifelse(min_lon < 0, paste0(abs(min_lon), "W"), paste0(min_lon, "E")), ")", "/",
         "VALUES", "/",
         "Y", "/",
         "(", ifelse(min_lat < 0, paste0(abs(min_lat), "S"), paste0(min_lat, "N")), ")", "/",
         "VALUES", "/")
}

add_t_range <- function(path, min_date, max_date) {
  paste0(path, "T", "/",
         "(", lubridate::day(min_date), "%20", lubridate::month(min_date, label = TRUE), 
         "%20", lubridate::year(min_date), ")", "/",
         "(", lubridate::day(max_date), "%20", lubridate::month(max_date, label = TRUE), 
         "%20", lubridate::year(max_date), ")", "/",
         "RANGEEDGES", "/"
  )
}

add_nc <- function(path) {
  paste0(path, "data.nc")
}
```
```
DataBook$set("public", "download_from_IRI", function(source, data, path = tempdir(), min_lon, max_lon, min_lat, max_lat, min_date, max_date, name, download_type = "point", import = TRUE){
  init_URL <- "https://iridl.ldeo.columbia.edu/"
  if(source == "CHIRPS_V2P0"){prexyaddress <- paste0(init_URL, "SOURCES/.UCSB/.CHIRPS/.v2p0")
  if(data == "daily_0p05") {extension <- ".daily/.global/.0p05/.prcp"}
  else if(data == "daily_0p25") {extension <- ".daily/.global/.0p25/.prcp"}
  else if(data == "daily_improved_0p05") {extension <- ".daily-improved/.global/.0p05/.prcp"}
  else if(data == "daily_improved_0p25") {extension <- ".daily-improved/.global/.0p25/.prcp"}
  else if(data == "dekad") {extension <- ".dekad/.prcp"}
  else if(data == "monthly_c8113") {extension <- ".monthly/.global/.c8113/.precipitation"}
  else if(data == "monthly_deg1p0") {extension <- ".monthly/.global/.deg1p0/.precipitation"}
  else if(data == "monthly_NMME_deg1p0") {extension <- ".monthly/.global/.NMME_deg1p0/.precipitation"}
  else if(data == "monthly_prcp") {extension <- ".monthly/.global/.precipitation"}
  else stop("Data file does not exist for CHIRPS V2P0 data")
  #Annual and 2Monthly and 3monthly does not exist for CHIRPS_V2P0
  }else if(source == "TAMSAT") {prexyaddress <-paste0(init_URL, "home/.remic/.Reading/.Meteorology/.TAMSAT")
  if(data == "rainfall_estimates") {extension <- ".TAMSAT-RFE/.rfe"}
  else if(data == "reconstructed_rainfall_anomaly") {extension <- ".TAMSAT-RFE/.rfediff"}
  else if(data == "sahel_dry_mask") {extension <- ".TAMSAT-RFE/.sahel_drymask"}
  else if(data == "SPI_1_dekad") {extension <- ".TAMSAT-RFE/.SPI-rfe_1-dekad_Sahel"}
  else stop("Data file does not exist for TAMSAT data")
  #monthly,climatology and TAMSAT RFE 0p1 are yet to be implemented.
  }else if(source == "NOAA_ARC2") {prexyaddress <- paste0(init_URL, "SOURCES/.NOAA/.NCEP/.CPC/.FEWS/.Africa/.DAILY/.ARC2")
  if(data == "daily_estimated_prcp") {extension <- ".daily/.est_prcp"}
  else if(data == "monthly_average_estimated_prcp") {extension <- ".monthly/.est_prcp"}
  else stop("Data file does not exist for NOAA ARC2 data")
  }else if(source=="NOAA_RFE2") {prexyaddress <- paste0(init_URL, "SOURCES/.NOAA/.NCEP/.CPC/.FEWS/.Africa")
  if(data == "daily_estimated_prcp"){extension <- ".DAILY/.RFEv2/.est_prcp"}
  else stop("Data file does not exist for NOAA RFE2 data")
  }else if(source=="NOAA_CMORPH_DAILY" || source=="NOAA_CMORPH_3HOURLY" || source=="NOAA_CMORPH_DAILY_CALCULATED") {
    if(source=="NOAA_CMORPH_DAILY") {prexyaddress <- paste0(init_URL, "SOURCES/.NOAA/.NCEP/.CPC/.CMORPH/.daily")}
    else if(source == "NOAA_CMORPH_3HOURLY") {prexyaddress <- paste0(init_URL, "SOURCES/.NOAA/.NCEP/.CPC/.CMORPH/.3-hourly")}
    if(source == "NOAA_CMORPH_DAILY_CALCULATED") {prexyaddress <- paste0(init_URL, "SOURCES/.NOAA/.NCEP/.CPC/.CMORPH/.daily_calculated")}
    if(data == "mean_microwave_only_est_prcp") {extension <- ".mean/.microwave-only/.comb"}
    else if(data == "mean_morphed_est_prcp") {extension <- ".mean/.morphed/.cmorph"}
    if(data == "orignames_mean_microwave_only_est_prcp") {extension <- ".orignames/.mean/.microwave-only/.comb"}
    else if(data == "orignames_mean_morphed_est_prcp") {extension <- ".orignames/.mean/.morphed/.cmorph"}
    if(data == "renamed102015_mean_microwave_only_est_prcp") {extension <- ".renamed102015/.mean/.microwave-only/.comb"}
    else if(data == "renamed102015_mean_morphed_est_prcp") {extension <- ".renamed102015/.mean/.morphed/.cmorph"}
    else stop("Data file does not exist for NOAA CMORPH data")
  }else if(source=="NASA_TRMM_3B42") {prexyaddress <- paste0(init_URL, "SOURCES/.NASA/.GES-DAAC/.TRMM_L3/.TRMM_3B42/.v7")
  if(data == "daily_estimated_prcp") {extension <- ".daily/.precipitation"}
  else if(data == "3_hourly_estimated_prcp") {extension <- ".three-hourly/.precipitation"}
  else if(data == "3_hourly_pre_gauge_adjusted_infrared_est_prcp") {extension <- ".three-hourly/.IRprecipitation"}
  else if(data == "3_hourly_pre_gauge_adjusted_microwave_est_prcp") {extension <- ".three-hourly/.HQprecipitation"}
  else stop("Data file does not exist for NASA TRMM 3B42 data")
  }else{stop("Source not specified correctly.")}
  prexyaddress <- paste(prexyaddress, extension, sep="/")
  if(download_type == "area"){xy_URL <- add_xy_area_range(path = prexyaddress, min_lon = min_lon, min_lat = min_lat, max_lon = max_lon, max_lat = max_lat)}
  else if(download_type == "point"){xy_URL <- add_xy_point_range(path = prexyaddress, min_lon = min_lon, min_lat = min_lat)}
  URL <- xy_URL %>% add_t_range(min_date = min_date, max_date = max_date) %>% add_nc()
  file_name <- tempfile(name, tmpdir = path, fileext = ".nc")
  result <- download.file(url = URL, destfile =  file_name, method = "auto", mode = "wb", cacheOK = FALSE)
  if(import & result == 0) {
    nc <- ncdf4::nc_open(filename = file_name)
    self$import_NetCDF(nc = nc, name = name)
    ncdf4::nc_close(nc=nc)
  }else if(result != 0){stop("No file downloaded please check your internet connection")}
  if(missing(path)) {file.remove(file_name)}
}
)
```

**Comment by shadrackkibet** [comment]
> We should consider which sources to include in the list. A range is good but I think too many gets confusing. We are mainly interested in the high resolution, long term rainfall products initially. Let's discuss what to include with Roger. I would like the names in the combo boxes to be more clearer as well, without _ etc.

@rdstern it will be good to get your thoughts on this. I now have something that works. So it will be good to sort out the list of sources.

**Comment by dannyparsons** [comment]
I've had a think about the design of the dialog which is as it was when it was first designed a while back. To make it more in line with how we now do dialogs I have a suggestion for a new design below. I don't think it's too much work to change. Does this look a bit better and clearer?

![iri_dialog](https://user-images.githubusercontent.com/11214914/93127551-0836fb80-f6c6-11ea-9e82-d6e674b180ba.jpg)

Below this, another change I have thought about after testing the dialog is to always keep the downloaded files. It now seems unhelpful to delete them after downloading, especially when it might take quite a lot of time. So can you change the "Save Downloaded File" checkbox to a "Save Downloaded Files To" label and change (again sorry!) the file selector back to a folder selector? And the default folder will be Downloads which VB has a function to easily get the path of. The file name could be something like this format: "Source_Data_timestamp.nc"

Keep the other checkbox not to import the data as it is.

**Comment by dannyparsons** [comment]
I've discussed the list of data with @rdstern to reorganise it. We suggest the Sources should mainly be the organisation list (except when there's clearly just one product that's well known). There might be more to add later but we focused on the current list.

So Sources and Data will be: 

**NASA**
Daily Precipitation
3-Hourly Precipitation

(I don't think we need the pre-gauge data as well)

**NOAA**
RFE2 Daily Precipitation
RFE2 10-day Precipitation

For CMORPH I don't quite understand all the different options. I think what we want is just one at each time interval:
- CMORPH 3-Hourly 
- CMORPH Daily
- CMORPH 5-day

But I don't quite understand all the current options and how they are different. So we need to find out what the sensible ones to include are. I also don't know how Daily and Daily-calculated are different, we might need both still but I'm not sure.

**TAMSAT**
v3.1 Daily Rainfall
v3.1 Daily Rainfall Complete (Filled)
v3.1 Dekadal Rainfall
v3.1 Dekadal Rainfall Complete (Filled)
v3.1 Monthly Rainfall
v3.1 Monthly Rainfall Complete (Filled)

**UCSB CHIRPS**
Daily Precipitation 0.05 degree
Daily Precipitation 0.25 degree
Dekad Precipitation 0.05 degree
Monthly Precipitation 0.05 degree

Let's start with this, and add more as we need. If we can't figure out CMORPH we can leave the code there but hide those options.

**Comment by rdstern** [comment]
As this is getting ready to be merged - and I am happy with that - want to check more (possibly once it has been merged) with what happens in practice.  I took 25 minutes for chirps for a single point and @dannyparsons says that's quite normal.  And that it may not be much longer if an area is downloaded.

But we should perhaps alert users to that - perhaps with a video included?  And what happens if the internet goes off during the downloading.  There is no change in the dialogue
